fixes #86: Reduce overhead of loading class definitions

guusdk opened the pull request:
This adds a simple cache that stores class definitions loaded by name. As these definitions are looked up for pretty much every serialization action in the cluster, these add significant overhead.

Classes could be loaded/unloaded dynamically. ~To account for that, the amount of time that class definitions survive in a cache should be limited.~

~This commit introduces two new properties:~
~- `hazelcast.cache.class.duration` - The maximum duration that a loaded class definition is cached.~
~- `hazelcast.cache.class.maxsize` - The maximum amount of cached class definitions (-1 for unlimited)~

Update: now listens to plugin changes

guusdk commented:
A run (very) roughly comparable with the one that was used to generate the statistics shown in the screenshots of #86 

![image](https://user-images.githubusercontent.com/4253898/196776191-96582098-2b84-4ffe-bf62-99109e47afec.png)
![image](https://user-images.githubusercontent.com/4253898/196775930-eed7d360-34c9-47f9-9fb4-6c2a2adb1b7f.png)

guusdk commented:
That one remaining hotspot curiously seems to have a single source. I wonder if there's room for improvement there.

![image](https://user-images.githubusercontent.com/4253898/196777082-e9807626-4e2d-4000-9e80-4d61a70d7ac7.png)

This was using a two-node cluster, based on Openfire 4.7.4 Alpha, build 4f9fa57

GregDThomas commented on src/java/org/jivesoftware/openfire/plugin/util/cache/ClusterClassLoader.java:
```diff
@@ -61,10 +65,24 @@ public class ClusterClassLoader extends ClassLoader {
 
     private final PluginClassLoader hazelcastClassloader;
 
+    // TODO: unload classes provided by plugins that are unloaded. Until then, keep the expiry very short, as to evict classes soon after unloading a plugin.
```
My normal comment; if there's a TODO, at least raise a ticket to cover it off later, and reference that ticket in the comment.

GregDThomas replied:
I don't know that a clusterable cache is a good idea. Even if a class can be serialized, I doubt that it would ever work on any other node than the node it was serialized on. Classpaths can be different, leading to subtly different classes being loaded (e.g. in a mixed-version cluster).

guusdk replied:
Oh, I agree. We have `CacheFactory.createLocalCache()` that I was planning to use for this. The implementation will be `org.jivesoftware.util.cache.DefaultLocalCacheStrategy`, which has the benefit of implementing Openfire's `Cache` interface (and thus show up in the cache summary pages, etc).

GregDThomas reviewed (approved):
Nice. One trivial comment, but LGTM

guusdk commented:
In code, there is no guarantee that the clustered class loader is used (but maybe, in practice, there is?).

I'm wondering if this should be moved from the `ClusterClassLoader` implementation to `org.jivesoftware.openfire.plugin.util.cache.ClusterExternalizableUtil#loadClass(java.lang.ClassLoader, java.lang.String)`.

GregDThomas commented:
I'd be very tempted to do that under a different ticket. The code, as-is, works. I can't see why moving it would make much difference, but may have side-effects that I can't see.

(i.e. is it there for a good reason that I can't see right now, or because of a bad design decision?)

guusdk commented:
> is it there for a good reason that I can't see right now

It is there right now because I've added it there in this PR. I'm not talking about moving the pre-existing ClusteredClassLoader implementation - merely the cache that I am adding for #86.

I have now added it to the ClusteredClassLoader, but perhaps I should have added it to the consumer of that implementation?

GregDThomas commented:
> I'm not talking about moving the pre-existing ClusteredClassLoader implementation - merely the cache that I am adding for #86.

Ah, sorry, misunderstood. I'd probably keep it where it is; that way there is one place where a plugin class is retrieved from; it may be from the cache, it may be from the classpath, but it's all in one place.

GregDThomas commented:
>  _only_ benefit this class loader. `org.jivesoftware.openfire.plugin.util.cache.ClusterExternalizableUtil#loadClass(java.lang.ClassLoader, java.lang.String)` can and will use others (notably, the Hazelcast-provided one), which would continue to cause poor performance.

I thought there was only one place we did this, but my memory on this is so hazy now I'm more than happy to take your word for it! In which case, re-using the cache does make sense!

guusdk commented:
I have made further changes:
- The cache was moved from the ClusteredClassLoader (one of the things that loads classes) to the code that uses the classes that are loaded. The benefit of this is that there is a performance increase, no matter what class loader is used.
- Switched from the Caffeine cache implementation to Openfire's "local" cache, which removes the need for specific properties to configure the cache (instead, Openfire default properties for caches will work)
- Added a plugin listener that will clear the cache when a plugin gets unloaded. This removes the need for very short cache durations.

GregDThomas commented on src/java/org/jivesoftware/openfire/plugin/HazelcastPlugin.java:
```diff
@@ -55,6 +58,22 @@ public void pluginsMonitored() {
+
+            @Override
+            public void pluginDestroyed(String s, Plugin plugin) {
+                // Although it is more performant to only purge the definitions loaded by the class loader of the related
```
✔

GregDThomas reviewed (approved):
LGTM. One exceedingly trivial comment you can feel free to ignore.

guusdk commented on pom.xml:
```diff
@@ -46,6 +46,12 @@
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.13.1</version>
```
Updated.
